[full]: share env path better between users

gengjiawen opened the pull request:
fix https://github.com/gitpod-io/workspace-images/issues/472

jankeromnes commented on full/Dockerfile:
```diff
@@ -233,4 +233,10 @@ RUN for i in $(ls /var/lib/apt/dazzle-marks/*.gpg); do apt-key add "$i"; done
 
+# share env see https://github.com/gitpod-io/workspace-images/issues/472
+USER gitpod
+RUN echo "PATH="${PATH}"" > /tmp/env.txt
+USER root
+RUN cat /tmp/env.txt >> /etc/environment && npx envinfo
```
Many thanks for implementing this fix! This is awesome.

Nit: I'd prefer not switching to `USER root` (it's not necessary), and not running `npx envinfo` in our build if possible.

```suggestion
RUN echo "PATH="${PATH}"" | sudo tee -a /etc/environment
```

jankeromnes reviewed (commented):
Hey, thank you for reporting this annoying problem, and for even fixing it! 💯

I'd love to have this merged. But first, please:

- Take a look at the in-line suggestion, and accept (or dismiss) it
- After this, please request that someone with write access triggers the CI for your PR using [these instructions](https://github.com/gitpod-io/workspace-images/blob/master/CONTRIBUTING.md#how-to-make-circleci-build-a-pr) (Note: I will be on holidays until 2021-09-16)

Then, when the `gitpod/workspace-full:branch-pr-474` image has been successfully built and deployed, please create a dummy repository with a `.gitpod.yml` that points to this temporary Docker image, and then open that repository in Gitpod, in order to test-drive the resulting image.

Finally, when the tests have confirmed that everything works, we can merge this Pull Request. (Sorry for making this a bit long and complicated 😬 the problem is that any change immediately goes live to every Gitpod user, so we need to be extra careful.)

Thanks again for your contributions! 👍

gengjiawen commented:
verified on my local machine, still works :)

jankeromnes commented:
Many thanks @gengjiawen!

Could you please already squash your commits now, and rebase on top of latest `master`? (This will allow triggering the CI from this PR in order to test that it also works with CircleCI)

gengjiawen commented:
@jankeromnes I can rebase. But Github support squash commit in UI, is it necessary to test it only after squashed it ?

jankeromnes commented:
> I can rebase.

Perfect, many thanks!

> But Github support squash commit in UI, is it necessary to test it only after squashed it ?

That's true, but for the CI to work on this Pull Request, we first needed to include a fix that recently landed in `master`: https://github.com/gitpod-io/workspace-images/commit/fdafb0c126e20548cd12190323c23cd9ba70b18a (otherwise the CI would have just failed for an unrelated reason, and we wouldn't have been able to test this PR before merging it)

gengjiawen commented:
rebased and squashed :)

On Mon, Sep 20, 2021 at 6:37 PM Jan Keromnes ***@***.***>
wrote:

> I can rebase.
>
> Perfect, many thanks!
>
> But Github support squash commit in UI, is it necessary to test it only
> after squashed it ?
>
> That's true, but for the CI to work on this Pull Request, we first needed
> to include a fix that recently landed in master: fdafb0c
> <https://github.com/gitpod-io/workspace-images/commit/fdafb0c126e20548cd12190323c23cd9ba70b18a>
> (otherwise the CI would have just failed for an unrelated reason, and we
> wouldn't have been able to test this PR before merging it)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitpod-io/workspace-images/pull/474#issuecomment-922809311>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4V5SESUG3M52CXPN5MZX3UC4FFLANCNFSM5CG6XYUQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jankeromnes commented on full/Dockerfile:
```diff
@@ -245,4 +245,7 @@ RUN for i in $(ls /var/lib/apt/dazzle-marks/*.gpg); do apt-key add "$i"; done
 COPY tests /var/lib/dazzle/tests
 
+# share env see https://github.com/gitpod-io/workspace-images/issues/472
+RUN echo "PATH="${PATH}"" | sudo tee /etc/environment 
```
Whoops, please remove the trailing whitespace here (sorry, our validation checks are a bit unhelpful 🙄):

```suggestion
RUN echo "PATH="${PATH}"" | sudo tee /etc/environment
```

gengjiawen replied:
fixed.

gengjiawen commented:
@jankeromnes Looks CI passed.

jankeromnes commented:
> @jankeromnes Looks CI passed.

Indeed! 🎉 Now we can do [this](https://github.com/gitpod-io/workspace-images/pull/474#pullrequestreview-746257406):

> Then, when the `gitpod/workspace-full:branch-pr-474` image has been successfully built and deployed, please create a dummy repository with a `.gitpod.yml` that points to this temporary Docker image, and then open that repository in Gitpod, in order to test-drive the resulting image.

Could you please set up such a test repository & `.gitpod.yml`?

Then we'll be able to open it in Gitpod, and use the Terminal to double-check `echo $PATH` (and then `sudo su` + `echo $PATH`), and compare that with another Gitpod workspace using the current default image.

gengjiawen commented:
> > @jankeromnes Looks CI passed.
> 
> Indeed! 🎉 Now we can do [this](https://github.com/gitpod-io/workspace-images/pull/474#pullrequestreview-746257406):
> 
> > Then, when the `gitpod/workspace-full:branch-pr-474` image has been successfully built and deployed, please create a dummy repository with a `.gitpod.yml` that points to this temporary Docker image, and then open that repository in Gitpod, in order to test-drive the resulting image.
> 
> Could you please set up such a test repository & `.gitpod.yml`?
> 
> Then we'll be able to open it in Gitpod, and use the Terminal to double-check `echo $PATH` (and then `sudo su` + `echo $PATH`), and compare that with another Gitpod workspace using the current default image.

https://github.com/gengjiawen/ts-scaffold/pull/309
```console
gitpod@ws-4dbf5d43-756c-4831-a584-ce1fbc84d716 /w/ts-scaffold (gengjiawen-patch-1)> echo $PATH
/home/gitpod/.sdkman/candidates/maven/current/bin /home/gitpod/.sdkman/candidates/java/current/bin /home/gitpod/.sdkman/candidates/gradle/current/bin /home/gitpod/.rvm/gems/ruby-2.7.3/bin /home/gitpod/.rvm/gems/ruby-2.7.3@global/bin /home/gitpod/.rvm/rubies/ruby-2.7.3/bin /home/gitpod/.pyenv/plugins/pyenv-virtualenv/shims /workspace/.cargo/bin /workspace/.pip-modules/bin /workspace/.rvm/bin /ide/bin /home/linuxbrew/.linuxbrew/bin /home/linuxbrew/.linuxbrew/sbin/ /workspace/.cargo/bin /workspace/.rvm/bin /workspace/.pip-modules/bin /home/gitpod/.pyenv/bin /home/gitpod/.pyenv/shims /workspace/go/bin /home/gitpod/go/bin /home/gitpod/go-packages/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /home/linuxbrew/.linuxbrew/bin /home/linuxbrew/.linuxbrew/sbin/ /home/gitpod/.nvm/versions/node/v14.17.6/bin /home/gitpod/.cargo/bin /usr/games /usr/games /home/gitpod/.nvm/versions/node/v14.17.6/bin /home/gitpod/.rvm/bin
gitpod@ws-4dbf5d43-756c-4831-a584-ce1fbc84d716 /w/ts-scaffold (gengjiawen-patch-1)> sudo su
root@ws-4dbf5d43-756c-4831-a584-ce1fbc84d716:/workspace/ts-scaffold# echo $PATH
/workspace/.cargo/bin:/workspace/.rvm/bin:/workspace/.pip-modules/bin:/home/gitpod/.pyenv/bin:/home/gitpod/.pyenv/shims:/workspace/go/bin:/home/gitpod/go/bin:/home/gitpod/go-packages/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin/:/home/gitpod/.nvm/versions/node/v14.17.6/bin:/home/gitpod/.cargo/bin:/usr/games:/usr/games
```

jankeromnes commented:
Many thanks for setting up a test branch! 🙏

Here is what I get (when comparing current production with this new change):

### User `gitpod`

```diff
  echo $PATH
- /home/gitpod/.sdkman/candidates/maven/current/bin /home/gitpod/.sdkman/candidates/java/current/bin /home/gitpod/.sdkman/candidates/gradle/current/bin /home/gitpod/.rvm/gems/ruby-2.7.3/bin /home/gitpod/.rvm/gems/ruby-2.7.3@global/bin /home/gitpod/.rvm/rubies/ruby-2.7.3/bin /home/gitpod/.pyenv/plugins/pyenv-virtualenv/shims /workspace/.cargo/bin /workspace/.pip-modules/bin /workspace/.rvm/bin /ide/bin /home/linuxbrew/.linuxbrew/bin /home/linuxbrew/.linuxbrew/sbin/ /workspace/.cargo/bin /workspace/.rvm/bin /workspace/.pip-modules/bin /home/gitpod/.pyenv/bin /home/gitpod/.pyenv/shims /workspace/go/bin /home/gitpod/go/bin /home/gitpod/go-packages/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /home/linuxbrew/.linuxbrew/bin /home/linuxbrew/.linuxbrew/sbin/ /home/gitpod/.nvm/versions/node/v14.17.5/bin /home/gitpod/.cargo/bin /usr/games /usr/games /home/gitpod/.nvm/versions/node/v14.17.5/bin /home/gitpod/.rvm/bin
+ /home/gitpod/.sdkman/candidates/maven/current/bin /home/gitpod/.sdkman/candidates/java/current/bin /home/gitpod/.sdkman/candidates/gradle/current/bin /home/gitpod/.rvm/gems/ruby-2.7.3/bin /home/gitpod/.rvm/gems/ruby-2.7.3@global/bin /home/gitpod/.rvm/rubies/ruby-2.7.3/bin /home/gitpod/.pyenv/plugins/pyenv-virtualenv/shims /workspace/.cargo/bin /workspace/.pip-modules/bin /workspace/.rvm/bin /ide/bin /home/linuxbrew/.linuxbrew/bin /home/linuxbrew/.linuxbrew/sbin/ /workspace/.cargo/bin /workspace/.rvm/bin /workspace/.pip-modules/bin /home/gitpod/.pyenv/bin /home/gitpod/.pyenv/shims /workspace/go/bin /home/gitpod/go/bin /home/gitpod/go-packages/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /home/linuxbrew/.linuxbrew/bin /home/linuxbrew/.linuxbrew/sbin/ /home/gitpod/.nvm/versions/node/v14.17.6/bin /home/gitpod/.cargo/bin /usr/games /usr/games /home/gitpod/.nvm/versions/node/v14.17.6/bin /home/gitpod/.rvm/bin
```

(no change)

### User `root`

```diff
  sudo su
  echo $PATH
- /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
+ /workspace/.cargo/bin:/workspace/.rvm/bin:/workspace/.pip-modules/bin:/home/gitpod/.pyenv/bin:/home/gitpod/.pyenv/shims:/workspace/go/bin:/home/gitpod/go/bin:/home/gitpod/go-packages/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin/:/home/gitpod/.nvm/versions/node/v14.17.6/bin:/home/gitpod/.cargo/bin:/usr/games:/usr/games
```

(more paths, but `/snap/bin`, `/usr/local/games` and `/usr/sbin` are removed)

---

However, I'm a bit surprised that it's still so different between `user` and `root`, so I compared both paths (using `echo $PATH | sed 's_[ :]_\n_g' | sort | uniq`, saving the output in two files, and then running `diff -up gitpod root`):

```diff
--- gitpod      2021-09-21 15:53:19.401051652 +0000
+++ root        2021-09-21 15:52:44.750807405 +0000
@@ -4,18 +4,9 @@
 /home/gitpod/go-packages/bin
 /home/gitpod/.nvm/versions/node/v14.17.6/bin
 /home/gitpod/.pyenv/bin
-/home/gitpod/.pyenv/plugins/pyenv-virtualenv/shims
 /home/gitpod/.pyenv/shims
-/home/gitpod/.rvm/bin
-/home/gitpod/.rvm/gems/ruby-2.7.3/bin
-/home/gitpod/.rvm/gems/ruby-2.7.3@global/bin
-/home/gitpod/.rvm/rubies/ruby-2.7.3/bin
-/home/gitpod/.sdkman/candidates/gradle/current/bin
-/home/gitpod/.sdkman/candidates/java/current/bin
-/home/gitpod/.sdkman/candidates/maven/current/bin
 /home/linuxbrew/.linuxbrew/bin
 /home/linuxbrew/.linuxbrew/sbin/
-/ide/bin
 /sbin
 /usr/bin
 /usr/games
```

So, it seems that user `root` is still missing a few paths that user `gitpod` has.

---

All in all, this seems to be better than before. @gengjiawen does this result work for you too? If so, I'd be happy to approve and merge this. 🙂

gengjiawen commented:
> All in all, this seems to be better than before. @gengjiawen does this result work for you too?

LGTM. Not sure why some are missing.
